Sources/OpenTelemetrySdk/Metrics/Export/PushMetricController.swift

```diff
@@ -6,35 +6,45 @@ import Foundation
 
 class PushMetricController {
     var meterSharedState: MeterSharedState
-    var meterProvider: MeterProviderSdk
+    unowned var meterProvider: MeterProviderSdk
 
     let pushMetricQueue = DispatchQueue(label: "org.opentelemetry.PushMetricController.pushMetricQueue")
+    let metricPushTimer: DispatchSourceTimer
 
     init(meterProvider: MeterProviderSdk, meterSharedState: MeterSharedState, shouldCancel: (() -> Bool)? = nil) {
         self.meterProvider = meterProvider
         self.meterSharedState = meterSharedState
-        pushMetricQueue.asyncAfter(deadline: .now() + meterSharedState.metricPushInterval) { [weak self] in
-            guard let self = self else {
-                return
-            }
-            while !(shouldCancel?() ?? false) {
-                autoreleasepool {
-                    let start = Date()
-                    let values = self.meterProvider.getMeters().values
-                    values.forEach {
-                        $0.collect()
-                    }
+        metricPushTimer = DispatchSource.makeTimerSource(flags: DispatchSource.TimerFlags(), queue: pushMetricQueue)
+        metricPushTimer.setEventHandler { [weak self] in
+            autoreleasepool {
+                guard let self = self else {
```

Comment:
```suggestion
                guard let self = self,
                      let meterProvider = self.meterProvider else {
```
Here we retain meterProvider by storing in a local variable so we can read all values without potential deletion